Add a --compile-extension command to build generated C extension

pablogsal opened the pull request:
To allow building the generated extesion when using the --cpython
command, add a --compile-extension command that automatically generates
an extension module (shared object). In this way, we can also build the
extensions easily on other platforms (like Linux).

gvanrossum commented on pegen/__main__.py:
```diff
@@ -8,12 +8,18 @@
 from __future__ import annotations  # Requires Python 3.7 or later
 
 import argparse
+from pathlib import Path
```
Just use `import pathlib` and write `pathlib.Path` twice.

gvanrossum commented:
And trying to add a `-o` flag just made it crash elsewhere:
```
$ python3 -m pegen -c --compile-extension data/cexpr.gram -o lala.c
Clean Grammar:
  start[mod_ty]: a=(stmt)* $ { Module ( a , NULL , p -> arena ) }
  stmt[stmt_ty]: a=expr_stmt { a }
  expr_stmt[stmt_ty]: a=expr NEWLINE { _Py_Expr ( a , EXTRA ( a , a ) ) }
  expr[expr_ty]: (l=expr '+' r=term { _Py_BinOp ( l , Add , r , EXTRA ( l , r ) ) } | l=expr '-' r=term { _Py_BinOp ( l , Sub , r , EXTRA ( l , r ) ) } | t=term { t })
  term[expr_ty]: (l=term '*' r=factor { _Py_BinOp ( l , Mult , r , EXTRA ( l , r ) ) } | l=term '/' r=factor { _Py_BinOp ( l , Div , r , EXTRA ( l , r ) ) } | f=factor { f })
  factor[expr_ty]: ('(' e=expr ')' { e } | a=atom { a })
  atom[expr_ty]: (n=NAME { n } | n=NUMBER { n } | s=STRING { s })
/Users/guido/pegen/pegen/pegen.c:6:1: warning: unused function 'token_name' [-Wunused-function]
token_name(int type)
^
1 warning generated.
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/runpy.py", line 192, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.8/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/guido/pegen/pegen/__main__.py", line 174, in <module>
    main()
  File "/Users/guido/pegen/pegen/__main__.py", line 127, in main
    shutil.move(cmd.get_ext_fullpath(extension_name),
  File "/usr/local/lib/python3.8/shutil.py", line 739, in move
    raise Error("Destination path '%s' already exists" % real_dst)
shutil.Error: Destination path './lala.cpython-38-darwin.so' already exists
$
```

pablogsal commented:
> TypeError: expected str, bytes or os.PathLike object, not NoneType

Fixed this on the latest ammend.

> shutil.Error: Destination path './lala.cpython-38-darwin.so' already exists

This error was due to the fact that the file already existed (the extension was already generated) and was going to be overwritten. In the latest ammend I have changed the code to always overwrite.

pablogsal commented:
I have tested the latest version on a mac system and it works now (both with -o and without). In addition, I have also:

1. Use the new build code in the Makefile.
2. Added a simple test to the test suite to check the c parser over some test cases.
3. Some minor refactors and cleanups.

Regarding the makefile, I have left your hardcoded path for Python:

```
TOP = /usr/local
PYTHON = $(TOP)/bin/python3.8
```

but in the future we can use just "python" and invoke it from a 3.8 virtualenv or something (which is what I am doing in Linux).

gvanrossum commented:
Very cool. Can we add this to Travis-CI too? Also just noticed that this broke the mypy build -- can you fix that?

pablogsal commented:
> Can we add this to Travis-CI too?
We have a test testing indirectly the `compile_c_extension` function added in this PR (the one testing the C parser). Do you have in mind something else to add to Travis-CI? Maybe separate the Python and the C tests?

> Also just noticed that this broke the mypy build -- can you fix that?

Yep, will prepare a PR soon :)
